### PR TITLE
Update compiler_builtins to 0.1.114

### PR DESCRIPTION
**Pull request by nicholasbishop**
The `weak-intrinsics` feature was removed from compiler_builtins in https://github.com/rust-lang/compiler-builtins/pull/598, so dropped the `compiler-builtins-weak-intrinsics` feature from alloc/std/sysroot.

In https://github.com/rust-lang/compiler-builtins/pull/593, some builtins for f16/f128 were added. These don't work for all compiler backends, so add a `compiler-builtins-no-f16-f128` feature and disable it for cranelift and gcc.

### CODE REVIEW

File: library/std/Cargo.toml [hunk 1]
```diff
@@ -17,7 +17,7 @@ cfg-if = { version = "1.0", features = ['rustc-dep-of-std'] }
 panic_unwind = { path = "../panic_unwind", optional = true }
 panic_abort = { path = "../panic_abort" }
 core = { path = "../core", public = true }
-compiler_builtins = { version = "0.1.105" }
+compiler_builtins = { version = "0.1.112", features = ["no-f16-f128"] }
```
Comment (RalfJung) [inline library/std/Cargo.toml:20]: So if someone only builds `core`, they still get the feature? That seems problematic?

Comment (bjorn3) [reply library/std/Cargo.toml:20]: Please instead add a `compiler-builtins-no-f16-f128` feature to liballoc, libstd and sysroot which enables `no-f16-f128` and then enable this feature at https://github.com/rust-lang/rust/blob/b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c/compiler/rustc_codegen_cranelift/build_system/build_sysroot.rs#L275 and https://github.com/rust-lang/rust/blob/b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c/compiler/rustc_codegen_gcc/build_system/src/build.rs#L115 for cg_clif and cg_gcc respectively.

Comment (nicholasbishop) [reply library/std/Cargo.toml]: I've made this change and pushed. However, when I tested with `ENABLE_GCC_CODEGEN=1 src/ci/docker/run.sh x86_64-gnu-llvm-17` it still failed with a panic related to f16/128. I don't think it ran the code in `rust/compiler/rustc_codegen_gcc/build_system/src/build.rs`. Is there another step I'm missing?

**Comment by klensy** [comment]
There exist also https://github.com/rust-lang/rust/pull/124886

**Comment by nicholasbishop** [comment]
Was this closed intentionally?

**Comment by GuillaumeGomez** [comment]
Oh wow, sorry didn't realize, no it wasn't on purpose!

File: library/std/Cargo.toml (hunk 1 above)
Comment (GuillaumeGomez) [reply library/std/Cargo.toml]: Unfortunately the current situation is pretty bad. [A PR](https://github.com/rust-lang/rust/pull/124353) to improve things is open and currently waiting from infra team information. Maybe give it a try?

**Comment by tgross35** [comment]
Note that https://github.com/rust-lang/rust/pull/124886 does the same thing

**Comment by nicholasbishop** [comment]
Makes sense. Is the `compiler-builtins-no-f16-f128` feature still needed here? I'm unclear whether that mangling error was the only problem, or if it's still correct to turn off f16/f128 for gcc/cranelift.

**Comment by tgross35** [comment]
> It appears the CI failure is being caused by the bootstrap stage0 compiler. I guess it won't succeed until #123816 reaches stage0 then.

I'll ask on Zulip if we might be able to backport that to unblock these updates

**Comment by tgross35** [comment]
I don’t have r+

r? @amanieu probably

**Comment by tgross35** [comment]
I am surprised this is still failing, the fix was backported in https://github.com/rust-lang/rust/pull/125529. I guess beta releases don't happen automatically?

**Comment by Mark-Simulacrum** [comment]
Beta should get released automatically, but will need to get bumped in src/stage0 by running the bump tool (./x.py run src/tools/bump-stage0).

**Comment by tgross35** [comment]
```
error: internal compiler error[0m[0m[1m: compiler/rustc_codegen_llvm/src/abi.rs:126:22: unsupported float: Reg { kind: Float, size: Size(2 bytes) }
```

We aren't actually testing f16 and f128 codegen _through_ LLVM anywhere yet, only up to the LLVM IR tests, looks like this is just a symptom of not having any full run-pass tests. I have been waiting on the compiler-builtins update to add more tests since we don't even have the most basic arithmetic available on all platforms without it, but looks like I need to figure this out first.

**Comment by nicholasbishop** [comment]
I was able to reproduce the error locally with `DEPLOY=1 src/ci/docker/run.sh dist-loongarch64-linux`. Added a 2nd commit to fill in f16/f128 in a couple places, and now dist-loongarch64-linux succeeds for me.

**Comment by tgross35** [comment]
Thanks! I'll still plan to add a test, but we can probably try this again since it's ready

@bors rollup=iffy I'm a bit worried about LLVM selection errors

r? @compiler-errors since the first commit touches your area

**Comment by compiler-errors** [comment]
Looks fine to me?

@bors r+ rollup=never

**Comment by nicholasbishop** [comment]
Well that's weird, I thought I successfully ran `DEPLOY=1 src/ci/docker/run.sh dist-loongarch64-linux` as described above. Must have misread the output or something though, since I can repro the latest failure:
```
rustc-LLVM ERROR: Cannot select: 0x7fc24c152d80: i64 = fp_to_fp16 0x7fc24c10f3c0
  0x7fc24c10f3c0: f32,ch = CopyFromReg 0x7fc24c0643b0, Register:f32 %5
    0x7fc24c10f900: f32 = Register %5
In function: _ZN17compiler_builtins5float6extend13__extendhfsf217ha6f6fd8c57385fa9E
[RUSTC-TIMING] memchr test:false 0.457
[RUSTC-TIMING] compiler_builtins test:false 1.395
rustc exited with signal: 4 (SIGILL) (core dumped)
```

**Comment by tgross35** [comment]
It's a LLVM bug https://github.com/llvm/llvm-project/issues/93894, LLVM's support for the new float types has turned out to vary quite a bit across targets.

Unfortunately that likely won't be fixed for a while, so probably the best we can do is to move compiler-builtins to a [platform-specific dependency](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies). Just add the `no-f16-f128` feature for `cfg(target_arch = "loongarch64")`.

There may be other failures so if you aren't up for more hunting around, feel free to just enable `no-f16-f128` for everything and I'll figure it out later.

**Comment by nicholasbishop** [comment]
Added the loongarch dep in `library/std/Cargo.toml`: https://github.com/rust-lang/rust/pull/125016/files#diff-bb1aff09a7d3bea25a8ed1520fea9dc273702fd4b772e45f5c4e584391037adeR42-R43

Is that the right place for it? (`dist-loongarch64-linux` does pass locally for me with that change.)

**Comment by bjorn3** [comment]
I think adding it to liballoc instead makes more sense. That way no_std loongarch targets disable it too.

**Comment by nicholasbishop** [comment]
Moved it to liballoc.

**Comment by tgross35** [comment]
Loongarch got fixed in LLVM but it will be a while before we get that. However, I just happened to try building something with f16 on s390x and it has a similar selection issue. (I'll open the llvm issues after I test more targets)

Maybe just change it so only x86 and aarch64 have f16/f128 and it's disabled on all others. Or still okay to disable everything if you prefer.

**Comment by nicholasbishop** [comment]
Updated to:
```toml
[target.'cfg(not(any(target_arch = "aarch64", target_arch = "x86", target_arch = "x86_64")))'.dependencies]
compiler_builtins = { version = "0.1.112", features = ["no-f16-f128"] }
```

**Comment by tgross35** [comment]
I think this is ready for a retry if anyone can r+

**Comment by tgross35** [comment]
Seems like a weird place for a failure. The tests do rely on both int and float math https://github.com/rust-lang/rust/blob/11380368dc53d0b2fc3a627408818eff1973ce9a/library/core/tests/num/int_log.rs#L102 (wish they printed the `i` so we could tell which half is wrong). I wonder if the change to remove weak linkage means one of our functions that has a bug is now getting linked?

Nothing else from https://github.com/rust-lang/compiler-builtins/compare/0.1.109...0.1.112 seems like it could be relevant.

```
2024-06-17T19:08:09.5367686Z ---- num::int_log::checked_ilog10 stdout ----
2024-06-17T19:08:09.5368598Z thread 'num::int_log::checked_ilog10' panicked at library/core/tests/num/int_log.rs:102:9:
2024-06-17T19:08:09.5369318Z assertion `left == right` failed
2024-06-17T19:08:09.5369710Z   left: Some(2)
2024-06-17T19:08:09.5370011Z  right: Some(1)
2024-06-17T19:08:09.5370199Z 
2024-06-17T19:08:09.5370205Z 
2024-06-17T19:08:09.5370321Z failures:
2024-06-17T19:08:09.5370998Z     num::int_log::checked_ilog10
```

**Comment by tgross35** [comment]
Was the symbols in compiler_builtins/libm strong in the previous version then?

**Comment by tgross35** [comment]
Assuming that picking the system symbols is sufficient to make this work, here is a PR to just disable our symbols on affected platforms https://github.com/rust-lang/compiler-builtins/pull/628

**Comment by fmease** [comment]
sync @bors r-

**Comment by tgross35** [comment]
@nicholasbishop 0.1.113 was released which should hopefully fix that issue. Could you update the version? https://github.com/rust-lang/compiler-builtins/pull/635#event-13288340054

**Comment by tgross35** [comment]
Entire CI is done... except arm-android is stuck between `zero-sized-tuple-struct.rs` and `zero-sized-btreemap-insert.rs`.

**Comment by matthiaskrgr** [comment]
-.-

@bors r=amanieu rollup=maybe

**Comment by klensy** [comment]
Need to update title to 0.1.113.

**Comment by tgross35** [comment]
It looks like this update is causing the arm-android CI test to get stuck, https://github.com/rust-lang-ci/rust/actions/runs/9687049474/job/26730826305, same thing happened yesterday. Seems like it executes fine until it is time to run `mpsc_stress.rs`, and then it can't continue.

I wonder if the weak linkage change is affecting something, seems like a weird failure mode.

In case there's something easy here, @rustbot ping android

**Comment by tgross35** [comment]
😕 @maurer @mgeisler then

**Comment by tgross35** [comment]
@nicholasbishop there is another new release of builtins that should hopefully fix the android issue, 0.1.114. Could you update this?

**Comment by tgross35** [comment]
Thanks for the quick change, hopefully this one works.

@bors r+

**Comment by TimNN** [comment]
I think something in this PR might be broken with regard to whether the f16/f128 builtins are being generated.

I have downloaded https://ci-artifacts.rust-lang.org/rustc-builds/80d8270d8488957f62fbf0df7a19dfe596be92ac/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz

And ran

```
llvm-nm ./rust-std-nightly-x86_64-unknown-linux-gnu/rust-std-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-7586cb84705120bb.rlib | grep __extend
./rust-std-nightly-x86_64-unknown-linux-gnu/rust-std-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-7586cb84705120bb.rlib:lib.rmeta: no symbols
0000000000000000 T _ZN17compiler_builtins5float6extend13__extendsfdf217hc5379a65d797e83bE
0000000000000000 W __extendsfdf2
```

I would have expected to see the f16 & f128 symbols there.

Running a `rustc` build in verbose mode, I see:

`/Users/logic/Projects/rustc/llvm-head/build/bootstrap/debug/rustc /Users/logic/Projects/rustc/llvm-head/build/bootstrap/debug/rustc --crate-name compiler_builtins --edition=2021 /Users/logic/.cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.114/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=145 --crate-type lib --emit=dep-info,metadata,link  --cfg 'feature="compiler-builtins"' --cfg 'feature="core"' --cfg 'feature="default"' --cfg 'feature="no-f16-f128"' --cfg 'feature="rustc-dep-of-std"' --target aarch64-apple-darwin  <truncated>`

Note the `--cfg 'feature="no-f16-f128"'` in there. This is on `aarch64` which means that, AFAICT, `no-f16-f128` should **not** trigger.

Commenting out

```toml
[target.'cfg(not(any(target_arch = "aarch64", target_arch = "x86", target_arch = "x86_64")))'.dependencies]
compiler_builtins = { version = "0.1.114", features = ["no-f16-f128"] }
```

Does produce a `libcompiler_builtins.rlib` which has the f16/f128 symbols.

**Comment by tgross35** [comment]
@TimNN thanks for reporting that, asking about it on Zulip https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Enabling.20features.20on.20specific.20targets.

**Comment by tgross35** [comment]
@TimNN since you have this pulled already, could you try what @epage suggested here? https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Enabling.20features.20on.20specific.20targets/near/454917731

**Comment by tgross35** [comment]
Actually let's move discussion to a new issue, opened https://github.com/rust-lang/rust/issues/128358.

**Comment by aeubanks** [comment]
This has also exposed https://github.com/rust-lang/compiler-builtins/issues/653

**Pull request merged**